_posts/2019-08-08-hacking-for-cruise-control.md

```diff
@@ -0,0 +1,270 @@
+---
+layout: post
+title:  "Hacking Strimzi for Cruise Control"
+date: 2019-08-08
+author: kyle_liberti
+---
+
+Cruise Control is a highly customizable software tool for simplifying the monitoring and balancing of workloads across a Kafka cluster.
+The flexibility of Cruise Control makes it easy to customize for different environments and use cases.
+This makes it possible to extend Cruise Control to work with Strimzi in an Kubernetes environment with a reasonable amount of effort.
+This post will walk through how to hook Cruise Control up to a Strimzi deployed Kafka cluster.
+<!--more-->
+
+# Prerequisite Outline
+
+Like any proof of concept, there are a few considerations and tradeoffs to think about before jumping in. For starters, one will need a Kafka cluster. At this time of this post, Cruise Control officially supports Kafka version 2.0.0 but should still be compatible with later versions of Kafka. This blog post will stick to using Kafka version 2.3.0 but feel free to experiment with other Kafka versions.
+
+The core challenges of hacking Cruise Control to work with Strimzi are: 
+
+- Gathering Kafka Metrics for Cruise Control consumption
+- Enabling communication between Cruise Control and Zookeeper
+- Deploying Cruise Control in Kubernetes
+
+There are many ways of approaching these hurdles but this post will attempt to address these challenges with as few steps as possible.
+
+# Gathering Metrics
+In order to estimate partition activity and resources, Cruise Control needs a steady source of metric data from Kafka. Fortunately, Kafka tracks all sorts of metrics and makes them all accessible through pluggable components known as 'metric reporters'. Kafka metric reporters designate what, where, when, and how metrics are prepared for export. 
+
+Kafka maintains two core types of metrics:
+* **Yammer metrics** For _server-side_ metrics related to brokers. These metric objects are kept in a globally accessible Yammer metric registry which can be reached directly by metric reporters. 
+* **Kafka metrics** For _client-side_ metrics related to Kafka producers, consumers, connect, and streams. These metric objects are tracked in concurrent hash map and explicitly passed to metric reporters when created. 
+
+The type of metrics exposed by the metric reporters, whether it be Yammer metrics, Kafka metrics, or both, is solely dependent on the implementation of the metric reporter. For example, the JMX metric reporter, a metric reporter provided and automatically enabled by Kafka, only wraps **Kafka metrics** as JMX MBeans. This is counterintuitive since it's known that Kafka exposes all of its metrics, Kafka *and* Yammer, through JMX. The reason why the Kafka JMX metric reporter doesn't wrap **Yammer metrics** as MBeans is because they are already exposed as JMX MBeans by Kafka by default.
```

Comment:
I didn't understand this sentence after reading it, and I'm familiar with how metrics work in Kafka, so I think this needs work (assuming it's relevant enough to keep). But for me all you need is the next paragraph.